### PR TITLE
Fix logic of memory_cleanse() on MSVC and clean up docs

### PR DESCRIPTION
**Pull request by real-or-random**
When working on https://github.com/bitcoin-core/secp256k1/issues/185, I noticed that the logic in memory_cleanse(), which is supposed to clear memory securely, is weird on MSVC. While it's correct, it's at least a code smell because the code clears the memory twice on MSVC. This weirdness was introduced by #11558. 

This PR fixes the logic on MSVC and also improves the docs around this function. Best reviewed in individual commits, see the commit messages for more rationale. The second commit touches only comments.

### CODE REVIEW

**Comment by laanwj** [comment]
This is one of the things that has been 'fixed' so many times, can we please be sure it's okay this time?

**Comment by real-or-random** [comment]
Yeah, so there is no 100% optimal solution for clearing memory, but please note that the goal of this PR is not to start this discussion about what the right method is. This merely fixes a weirdness on in the logic introduced by #11558.

**Comment by practicalswift** [comment]
utACK 751aff1e50ec75f1f57dd72640eaca9a22ee7406 (read code - looks correct)

**Comment by MarcoFalke** [comment]
Fine on changing the docs, but is it really necessary to remove the redundant `std::memset(ptr, 0, len)`.

**Comment by real-or-random** [comment]
Well, as you say, the call is redundant, so it's okay not remove it it but I'd argue it's desired to remove it
 * for readability: It just looks wrong to do the same operation twice. I think people would complain about code that does `x = 0; x = 0;`.
 * for efficiency: While MSVC inlines the call to `SecureZeroMemory`, it is not clever enough to optimize the call to `memset` out, see https://godbolt.org/z/Y2TIWC.

File: src/support/cleanse.cpp [hunk 1]
```diff
@@ -11,33 +11,25 @@
...
+    std::memset(ptr, 0, len);
+
+    /* Memory barrier that scares the compiler away from optimizing out the memset.
+     *
+     * Quoting Adam Langley <agl@google.com> in commit ad1907fe73334d6c696c8539646c21b11178f20f
+     * in BoringSSL (ISC License):
+     *    As best as we can tell, this is sufficient to break any optimisations that
+     *    might try to eliminate "superfluous" memsets.
+     * This method used in memzero_explicit() the Linux kernel, too. Its advantage is that it is
+     * pretty efficient, because the compiler can still implement the memset() efficently,
```
Comment (practicalswift) [inline src/support/cleanse.cpp]: Should be "efficiently" :-)

Comment (real-or-random) [reply src/support/cleanse.cpp]: fixed with force-push

**Comment by practicalswift** [comment]
ACK 4e1941e8bd3a72adb2291af8148bee8a508bcd95 (read code - looks correct)

File: src/support/cleanse.cpp [hunk 2]
```diff
@@ -11,33 +11,25 @@
...
-    /* As best as we can tell, this is sufficient to break any optimisations that
-       might try to eliminate "superfluous" memsets. If there's an easy way to
-       detect memset_s, it would be better to use that. */
+    /* Memory barrier that scares the compiler away from optimizing out the memset.
+     *
+     * Quoting Adam Langley <agl@google.com> in commit ad1907fe73334d6c696c8539646c21b11178f20f
+     * in BoringSSL (ISC License):
+     *    As best as we can tell, this is sufficient to break any optimisations that
+     *    might try to eliminate "superfluous" memsets.
+     * This method used in memzero_explicit() the Linux kernel, too. Its advantage is that it is
```
Comment (jonasnick) [inline src/support/cleanse.cpp]: nit: I can't parse the first sentence - perhaps missing a verb and "in"?

**Review by jonasnick — APPROVED** [review]
ACK read the code and the comments. Comments are way better than before. The cited paper ("Dead Store Elimination (Still) Considered Harmful") is an excellent resource.

**Comment by practicalswift** [comment]
@real-or-random Would you mind fixing the typo? After that I think this should be ready for merge :-)

**Comment by real-or-random** [comment]
@practicalswift @jonasnick 
 I fixed the typo, can you reACK?

**Comment by practicalswift** [comment]
utACK f53a70ce95231d34bb14cd6c58503927e8d7ff59 :-)

**Comment by laanwj** [comment]
code review ACK f53a70ce95231d34bb14cd6c58503927e8d7ff59

**Comment by laanwj** [comment]
Seeing the merge message this PR title is badly worded: yes, you can consider clearing the memory twice a logic error, but most people would consider it a logic error if it would result in *not clearing the memory*.

So to be 100% clear for anyone getting here: this is not a security fix.

**Pull request merged**